src/pybind/mgr/volumes/module.py

```diff
@@ -353,6 +352,11 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
             'desc': "Fetch a export of a NFS cluster given the pseudo path/binding",
             'perm': 'r'
         },
+        {
+            'cmd': 'nfs export update ',
+            'desc': "Update an export of a NFS cluster by `-i <json_file>`",
```

Comment:
Doesn't look like we actually parse the `-i` option at all -- we always ready from stdin? That's fine but maybe we should get rid of this option. (It's not clear to me what the file name would be good for since the mons are on another machine.)

Reply:
No, we actually parse through `-i` option. I am not sure if the cmd can be directly specified here as `nfs export update -i`

Reply:
Where is it parsed. AFAICT, the command dispatch drops the `cmd` dictionary on line 715.

Reply:
Here
https://github.com/ceph/ceph/blob/6a5dfea50a0d5e55a3f81c0242aabaa80c545ae9/src/pybind/mgr/volumes/fs/nfs.py#L813-L819

Reply:
There you're loading the config but the `cmd` argument to `_cmd_nfs_export_update` is dropped which holds the arguments (`<json_file>`).

Reply:
??

Reply:
I have dropped '-i'

nfs export update <jsonfile> :  Update an export of a NFS cluster

```
diff --git a/src/pybind/mgr/volumes/fs/nfs.py b/src/pybind/mgr/volumes/fs/nfs.py
index f47597a122d..ab6cdbf3011 100644
--- a/src/pybind/mgr/volumes/fs/nfs.py
+++ b/src/pybind/mgr/volumes/fs/nfs.py
@@ -814,6 +814,8 @@ class FSExport(object):
         try:
             if not export_config:
                 return -errno.EINVAL, "", "Empty Config!!"
+            with open(export_config) as exportfile:
+                export_config = exportfile.read()
             update_export = json.loads(export_config)
             old_export, update_user_caps = self._validate_export(update_export)
             if update_user_caps:
diff --git a/src/pybind/mgr/volumes/module.py b/src/pybind/mgr/volumes/module.py
index 3ad65861c3d..7b5336d0f57 100644
--- a/src/pybind/mgr/volumes/module.py
+++ b/src/pybind/mgr/volumes/module.py
@@ -353,8 +353,9 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
             'perm': 'r'
         },
         {
-            'cmd': 'nfs export update ',
-            'desc': "Update an export of a NFS cluster by `-i <json_file>`",
+            'cmd': 'nfs export update '
+                   'name=jsonfile,type=CephFilepath ',
+            'desc': "Update an export of a NFS cluster",
             'perm': 'rw'
         },
         {
@@ -712,7 +713,8 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
 
     @mgr_cmd_wrap
     def _cmd_nfs_export_update(self, inbuf, cmd):
-        return self.fs_export.update_export(export_config=inbuf)
+        #return self.fs_export.update_export(export_config=inbuf)
+        return self.fs_export.update_export(export_config=cmd['jsonfile'])
 
     @mgr_cmd_wrap
     def _cmd_nfs_cluster_create(self, inbuf, cmd):
```

But with this change all file related checks needs to be done.

Reply:
> There you're loading the config but the `cmd` argument to `_cmd_nfs_export_update` is dropped which holds the arguments (`<json_file>`).

`<json_file>` is not part of `cmd` argument. Rather it is passed as `inbuf`.

https://github.com/ceph/ceph/blob/d4292afebffe28648a2d27cbf1597b82a77c8593/src/ceph.in#L1136-L1147

https://github.com/ceph/ceph/blob/d4292afebffe28648a2d27cbf1597b82a77c8593/src/pybind/mgr/mgr_module.py#L1331-L1347

Reply:
Okay, so there is some "magic" for handling `-i` argument.

Reply:
Can you add a comment to `_cmd_nfs_export_update` noting that `-i` processing is done by the Ceph CLI.

Reply:
Done